File: DialedInCoffee/RulesEngine/Rule.swift
import Foundation

public struct Rule: Identifiable, Hashable, Sendable {
    public let id: String
    public let trigger: Symptom
    public let cause: Cause
    public let baseWeight: Double           // 0...1 base contribution when trigger fires
    public let conditions: [Condition]      // ALL must be satisfied to multiply contribution
    public let conditionMultiplier: Double  // applied when conditions match (default 1.0)
    public let rationale: String

    public init(
        id: String,
        trigger: Symptom,
        cause: Cause,
        baseWeight: Double,
        conditions: [Condition] = [],
        conditionMultiplier: Double = 1.0,
        rationale: String
    ) {
        self.id = id
        self.trigger = trigger
        self.cause = cause
        self.baseWeight = baseWeight
        self.conditions = conditions
        self.conditionMultiplier = conditionMultiplier
        self.rationale = rationale
    }

    public func effectiveWeight(
        current: MachineSettings,
        drink: DrinkType,
        milk: Milk,
        bean: BeanSnapshot?
    ) -> Double {
        if conditions.isEmpty { return baseWeight }
        let allMatch = conditions.allSatisfy {
            $0.isSatisfied(current: current, drink: drink, milk: milk, bean: bean)
        }
        return allMatch ? min(1.0, baseWeight * conditionMultiplier) : baseWeight
    }
}

File: android/core/src/main/kotlin/com/cawfee/domain/rules/Rule.kt
package com.cawfee.domain.rules

import com.cawfee.domain.model.BeanSnapshot
import com.cawfee.domain.model.Cause
import com.cawfee.domain.model.DrinkType
import com.cawfee.domain.model.MachineSettings
import com.cawfee.domain.model.Milk
import com.cawfee.domain.model.Symptom

/** A symptom→cause rule with optional conditional weighting. Ported from Rule.swift. */
data class Rule(
    val id: String,
    val trigger: Symptom,
    val cause: Cause,
    val baseWeight: Double,                 // 0..1 base contribution when trigger fires
    val conditions: List<Condition> = emptyList(),
    val conditionMultiplier: Double = 1.0,  // applied when ALL conditions match
    val rationale: String,
) {
    fun effectiveWeight(current: MachineSettings, drink: DrinkType, milk: Milk, bean: BeanSnapshot?): Double {
        if (conditions.isEmpty()) return baseWeight
        val allMatch = conditions.all { it.isSatisfied(current, drink, milk, bean) }
        return if (allMatch) minOf(1.0, baseWeight * conditionMultiplier) else baseWeight
    }
}
